src/passes/OptimizeInstructions.cpp

```diff
@@ -1949,12 +1925,27 @@ struct OptimizeInstructions
     //
     // The fallthrough is the local.get, but the array cast in the middle
     // proves a trap must happen.
+    Builder builder(*getModule());
+    auto nullType = curr->type.getHeapType().getBottom();
     {
       auto* ref = curr->ref;
       while (1) {
         auto result = GCTypeUtils::evaluateCastCheck(ref->type, curr->type);
 
-        if (result == GCTypeUtils::Failure) {
+        if (result == GCTypeUtils::Success) {
+          // The cast will succeed, but we can't just remove the cast and
+          // replace it with `ref` because the intermediate expressions might
+          // have had side effects. If we know the value is a null, though, we
+          // can at least put the side effects in a block that directly returns
+          // the null value.
+          // TODO: Track side effects of expressions we've seen to determine
+          // whether we can just replace the cast with `ref`.
```

Comment:
I think we don't need this TODO as vacuum will remove the dropped ref, or parts of it, when possible. So we don't need to try to do better here.

Reply:
This TODO is to allow us to optimize things other than nulls where we wouldn't be able to just drop the child expression because we wouldn't have a value to return afterward.

Reply:
Ok, I think we can do that without tracking side effects - maybe that's what confused me. We can add a local.tee to stash the value, and return a local.get of it?

Reply:
Oh hmm, yes, that sounds like it would work. Are we worried about cases where it makes the code worse because there actually are side effects that vacuum wouldn't be able to get rid of?

Reply:
I'm not very worried about that. In trapsNeverHappen mode vacuum should almost always succeed. And the block vanishes anyhow. I suppose if we wanted to be careful we could gate this opt on trapsNeverHappen, but the benefits may lead to bigger savings later anyhow even without it.

GUFA does this sort of thing and it's not had downsides. Which makes me realize that this overlaps with GUFA as it tracks constant values like nulls through expressions and then emits them on the outside. But as this is quite simple it still seems worthwhile.

Reply:
Sounds good, thanks. I'll update the TODO and implement this in a follow-up.